crate_universe/src/metadata/metadata_annotation.rs

```diff
@@ -203,7 +203,7 @@ impl LockfileAnnotation {
         metadata: &CargoMetadata,
         lockfile: &CargoLockfile,
         workspace_metadata: &WorkspaceMetadata,
-    ) -> Result<SourceAnnotation> {
+    ) -> Result<Option<SourceAnnotation>> {
```

Comment:
Instead of having source annotations become optional, why not create a new variant `SourceAnnotation::Path`? This way a missing annotation could be interpreted as an error or lack of data.